### PR TITLE
Support glob pattern to exclude unknown validation

### LINKED ISSUE #79: Support glob pattern to exclude unknown validation
Use https://github.com/eclipse/m2e.wtp/blob/master/org.eclipse.m2e.wtp/src/org/eclipse/m2e/wtp/internal/AntPathMatcher.java

### PR DESCRIPTION
**Pull request by angelozerr**
Support glob pattern to exclude unknown validation

Fixes #79

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
@fbricon excluded pattern for unknown validation should work. It misses tests (I create the PR as draft)

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/AntPathMatcher.java [hunk 1]
```diff
@@ -0,0 +1,938 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * {@link PathMatcher} implementation for Ant-style path patterns.
+ *
+ * <p>Part of this mapping code has been kindly borrowed from <a href="https://ant.apache.org">Apache Ant</a>.
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/AntPathMatcher.java]: indicate latest provenance

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/PathMatcher.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+import java.util.Comparator;
+import java.util.Map;
+
+/**
+ * Strategy interface for {@code String}-based path matching.
+ *
+ * <p>Used by {@link org.springframework.core.io.support.PathMatchingResourcePatternResolver},
+ * {@link org.springframework.web.servlet.handler.AbstractUrlHandlerMapping},
+ * and {@link org.springframework.web.servlet.mvc.WebContentInterceptor}.
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/PathMatcher.java]: unnecessary comment

please indicate provenance from spring

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java [hunk 3]
```diff
@@ -95,12 +132,45 @@ private boolean isExcluded(String propertyName) {
 		if (excluded == null) {
 			return false;
 		}
-		for (String pattern : excluded) {
-			if (pattern.equals(propertyName)) {
+		// Get compiled excluded properties
+		List<ExcludedProperty> excludedProperties = getExcludedProperties();
+		for (ExcludedProperty excluded : excludedProperties) {
+			// the property name match an excluded pattern
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java]: matches

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java [hunk 4]
```diff
@@ -23,6 +28,38 @@
...
+		 * @param propertyName the property name.
+		 * @return true if the given property name match the pattern and false
+		 *         otherwise.
+		 */
+		public boolean match(String propertyName) {
+			if (matcher != null) {
+				// the excluded property is a pattern, use pattern matcher to check the match
+				return matcher.match(pattern, propertyName);
+			}
+			// the excluded property is not a pattern, check if the property name is equals
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java]: is equal

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java [hunk 5]
```diff
@@ -23,6 +28,38 @@
...
+		private final String pattern;
+		private final AntPathMatcher matcher;
+
+		public ExcludedProperty(String pattern, AntPathMatcher matcher) {
+			this.pattern = pattern;
+			this.matcher = matcher.isPattern(pattern) ? matcher : null;
+		}
+
+		/**
+		 * Returns true if the given property name match the pattern and false
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java]: matches

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java [hunk 6]
```diff
@@ -23,6 +28,38 @@
...
+		 * otherwise.
+		 * 
+		 * @param propertyName the property name.
+		 * @return true if the given property name match the pattern and false
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java]: matches

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java [hunk 7]
```diff
@@ -95,12 +132,45 @@ private boolean isExcluded(String propertyName) {
...
+		if (excludedProperties != null) {
+			return excludedProperties;
+		}
+		List<ExcludedProperty> excludedProperties = new ArrayList<>();
+		AntPathMatcher matcher = new AntPathMatcher();
+		matcher.setCachePatterns(true);
+		for (String pattern : excluded) {
+			excludedProperties.add(new ExcludedProperty(pattern, matcher));
+		}
+		return excludedProperties;
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java]: or 
```java
AntPathMatcher matcher = new AntPathMatcher();
matcher.setCachePatterns(true);
return Stream.of(excluded)
		.map(p -> new ExcludedProperty(p, matcher))
		.collect(Collectors.toList());
```

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/AntPathMatcher.java (hunk 1 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/AntPathMatcher.java]: @fbricon fixed

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/PathMatcher.java (hunk 2 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/PathMatcher.java]: @fbricon fixed

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java (hunk 4 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java]: @fbricon fixed

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java (hunk 7 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java]: Indeed it's better, @fbricon fixed

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java (hunk 3 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusValidationTypeSettings.java]: @fbricon fixed

**Comment by angelozerr** [comment]
Test has been written -> https://github.com/redhat-developer/quarkus-ls/pull/119/files#diff-82b7d3e3783d1daeb5f587fcb1baf3abR131

**Comment by angelozerr** [comment]
@fbricon please review in same time https://github.com/redhat-developer/vscode-quarkus/pull/118

**Comment by xorye** [comment]
I've tried the feature on my machine with this PR: https://github.com/redhat-developer/vscode-quarkus/pull/118, it is working great

**Comment by angelozerr** [comment]
> I've tried the feature on my machine with this PR: redhat-developer/vscode-quarkus#118, it is working great

Thank's @xorye for your feedback. The PR is just to add some description, could you review it please.

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/PathMatcher.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+import java.util.Comparator;
+import java.util.Map;
+
+/**
+ * Strategy interface for {@code String}-based path matching.
+ * 
+ * <p>
+ * This class is a copy/paste from
+ * https://github.com/spring-projects/spring-framework/blob/master/spring-core/src/main/java/org/springframework/util/PathMatcher.java
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/PathMatcher.java]: need exact version, link to sha1

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/AntPathMatcher.java [hunk 9]
```diff
@@ -0,0 +1,949 @@
...
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * {@link PathMatcher} implementation for Ant-style path patterns.
+ *
+ * <p>
+ * This class is a copy/paste from
+ * https://github.com/spring-projects/spring-framework/blob/master/spring-core/src/main/java/org/springframework/util/AntPathMatcher.java
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/AntPathMatcher.java]: need exact version, link to sha1

**Comment by fbricon** [comment]
ok to merge once the links to versioned files are fixed

**Pull request merged**